lib/vdsm/storage/hsm.py

```diff
@@ -2871,6 +2871,9 @@ def prune_bitmaps(self, sdUUID, imgUUID, volUUID, baseUUID):
             baseUUID (str): The UUID of the volume that will get its bitmaps
                 pruned if missing or invalid in top.
         """
+        vars.task.getSharedLock(STORAGE, imgUUID)
+        vars.task.getSharedLock(STORAGE, volUUID)
+        vars.task.getExclusiveLock(STORAGE, baseUUID)
         base_vol = self._produce_volume(sdUUID, imgUUID, baseUUID)
         top_vol = self._produce_volume(sdUUID, imgUUID, volUUID)
```

Comment:
This does not work, since we _produce_volume takes a shared lock on the volume, and this code tries to take the same lock twice:

    def _produce_volume(self, sdUUID, imgUUID, volUUID):
        vars.task.getSharedLock(STORAGE, sdUUID)
        ...

So code is very clean but incorrect. What we want to do is to have list of locks, and lock of of them at the same time and in the right order using guarded.context().

    img_ns = rm.getNamespace(sc.IMAGE_NAMESPACE, self.sd_id)
    locks = [rm.Lock(sc.STORAGE, sdUUID, rm.SHARED),
             rm.Lock(img_ns, imgUUID, rm.EXCLUSIVE)]
    with guarded.context(locks):
        prune bitmaps here...

You need to produce the volume manually instead of using the _produce_volume() helper, it does too much and cannot be used when using multiple volumes. Maybe it can be fixed to support this usage but lets keep this change simple and avoid unneeded changes that can break unrelated flows and need much more testing.